scratch/evaluate_scheduler_qd_dense_scenario.cc

```diff
@@ -0,0 +1,825 @@
+  info.SetAllocationType (SERVICE_PERIOD_ALLOCATION);
+  info.SetAllocationFormat (ISOCHRONOUS);
+  info.SetAsPseudoStatic (isPseudoStatic);
+  info.SetDestinationAid (AID_AP);
+  element.SetDmgAllocationInfo (info);
+  element.SetMinimumAllocation (minAllocation);
+  element.SetMaximumAllocation (maxAllocation);
+  element.SetMinimumDuration (minAllocation);
+  return element;
+}
+
+void
+StationAssociated (Ptr<Node> node, Ptr<DmgStaWifiMac> staWifiMac, Mac48Address address, uint16_t aid)
+{
+  NS_LOG_FUNCTION (node << staWifiMac << address << aid);
+  if (!csv)
+    {
+      NS_LOG_DEBUG ("DMG STA=" << staWifiMac->GetAddress () << " associated with DMG PCP/AP=" << address
+                    << ", AID=" << aid);
+    }
+  /* Send ADDTS request to the PCP/AP */
+  CommunicationPairListI it = communicationPairList.find (node);
+  if (it != communicationPairList.end ())
+    {
+      uint32_t spDuration = ComputeServicePeriodDuration (it->second.appDataRate, WifiMode (phyMode).GetPhyRate ());
+      staWifiMac->CreateAllocation (GetDmgTspecElement (allocationId++, true, spDuration, spDuration));
+    }
+  else
+    {
+      NS_FATAL_ERROR ("Could not find application for this node.");
+    }
+}
+
+void
+StationDeAssociated (Ptr<Node> node, Ptr<DmgWifiMac> staWifiMac, Mac48Address address)
+{
+  if (!csv)
+    {
+      NS_LOG_DEBUG ("DMG STA=" << staWifiMac->GetAddress () << " deassociated from DMG PCP/AP=" << address);
+    }
+  CommunicationPairListI it = communicationPairList.find (node);
+  if (it != communicationPairList.end ())
+    {
+      it->second.srcApp->StopApplication ();
+    }
+  else
+    {
+      NS_FATAL_ERROR ("Could not find application to delete.");
+    }
+}
+
+CommunicationPair
+InstallApplications (Ptr<Node> srcNode, Ptr<Node> dstNode, Ipv4Address address, string appDataRate, uint16_t appNumber)
+{
+  NS_LOG_FUNCTION (srcNode->GetId () << dstNode->GetId () << address << appDataRate << +appNumber);
+  CommunicationPair commPair;
+  /* Install TCP/UDP Transmitter on the source node */
+  Address dest (InetSocketAddress (address, 9000 + appNumber));
+  ApplicationContainer srcApp;
+  if (applicationType == "onoff")
+    {
+      OnOffHelper src (socketType, dest);
+      src.SetAttribute ("MaxBytes", UintegerValue (maxPackets));
+      src.SetAttribute ("PacketSize", UintegerValue (packetSize));
+      src.SetAttribute ("OnTime", StringValue ("ns3::ConstantRandomVariable[Constant=1e6]"));
+      src.SetAttribute ("OffTime", StringValue ("ns3::ConstantRandomVariable[Constant=0]"));
+      src.SetAttribute ("DataRate", DataRateValue (DataRate (appDataRate)));
+      srcApp = src.Install (srcNode);
+    }
+  else if (applicationType == "bulk")
+    {
+      BulkSendHelper src (socketType, dest);
+      srcApp = src.Install (srcNode);
+    }
+  /* The application is started when the corresponding ADDTS request succeeded (or failed only for CbapOnlyDmgWifiScheduler) */
+  srcApp.Start (Seconds (simulationTime + 1));
```

Comment:
is this needed?

Reply:
Yes, otherwise the app will start at 0 by default.

Reply:
ok, maybe improve the comment then, it is not clear enough IMO